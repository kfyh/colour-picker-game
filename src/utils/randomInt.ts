export async function randomInt(min: number, max: number): Promise<number> {
  const response = await fetch('https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8');
  if (response.ok) {
    const json = await response.json();
    const random = json.data[0] / 255;
    return Math.floor(min + random * (max - min));
  } else {
    throw new Error(response.statusText);
  }
}

export function randomIntLocal(min: number, max: number): number {
  const random = Math.random();
  return Math.floor(min + random * (max - min));
};
