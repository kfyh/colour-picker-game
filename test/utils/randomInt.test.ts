import { randomInt, randomIntLocal } from '../../src/utils/randomInt';
describe('randomInt', () => {
  describe('randomInt function', () => {
    let origFetch: (input: RequestInfo | URL, init?: RequestInit | undefined) => Promise<Response>;
    beforeEach(() => {
      origFetch = global.fetch;
    })
    afterEach(() => {
      global.fetch = origFetch;
    });
  
    test('result is integer', async () => {
      global.fetch = jest.fn().mockResolvedValue({ok: true, json: () => ({data: [123]})});
      
      const result = await randomInt(0, 4);
      expect(Number.isInteger(result)).toBeTruthy();
    });
  
    test('result is greater than min', async () => {
      global.fetch = jest.fn().mockResolvedValue({ok: true, json: () => ({data: [0]})});
  
      const result = await randomInt(0, 4);
      expect(result).toBeGreaterThanOrEqual(0);
    });
  
    test('result is less than max', async () => {
      global.fetch = jest.fn().mockResolvedValue({ok: true, json: () => ({data: [255]})});
  
      const result = await randomInt(0, 4);
      expect(result).toBeLessThanOrEqual(4);
    });

    test('Error when fetch fails', async () => {
      global.fetch = jest.fn().mockResolvedValue({ok: false, statusText: "error error"});
  
      await expect(async () => { await randomInt(0, 4)}).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('randomIntLocal', () => {
    test('result is integer', () => {
      for (let i = 0; i < 100; i++) {
        const result = randomIntLocal(0, 4);
        expect(Number.isInteger(result)).toBeTruthy();
      }
    });
  
    test('result is greater than min', () => {
      for (let i = 0; i < 100; i++) {
        const result = randomIntLocal(0, 4);
        expect(result).toBeGreaterThanOrEqual(0);
      }
    });
  
    test('result is less than max', () => {
      for (let i = 0; i < 100; i++) {
        const result = randomIntLocal(0, 4);
        expect(result).toBeLessThanOrEqual(4);
      }
    });
  })
});
