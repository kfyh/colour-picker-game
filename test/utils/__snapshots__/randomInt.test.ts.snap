// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`randomInt randomInt function Error when fetch fails 1`] = `"error error"`;
